### PR TITLE
sync_send_to implementation

### PR DESCRIPTION
**Pull request by sjrct**
I should write some tests for this, shouldn't be hard.

### CODE REVIEW

File: halibot/halobject.py [hunk 1]
```diff
@@ -48,6 +53,20 @@ def send_to(self, msg, dests):
...
+		futs = self.send_to(msg, dests)
+
+		r = {}
+		for name, fut in futs.items():
+			fut.result()
+			to = self._hal.objects.get(name)
+			if to and msg.uuid in to.sync_replies:
+				# Assure that the module was not removed in the interim
+				r[name] = to.sync_replies.pop(msg.uuid)
+		return r
```
Comment (richteer) [inline halibot/halobject.py:68]: So this returns a map of `module -> replies`, correct?
Will there ever be a case that we might want to fire off the `sync_` message, do some other things, then come back to waiting on that first message?

In other words, would it be worth returning the future so we can `await sync_send_to`?

Comment (richteer) [reply halibot/halobject.py:68]: Side note: probably not, just thinking out loud here

File: halibot/halobject.py [hunk 2]
```diff
@@ -3,6 +3,7 @@
 import inspect
 import copy
 from threading import Thread
+from collections import defaultdict
```
Comment (richteer) [inline halibot/halobject.py:6]: How did we not find this sooner

File: halibot/message.py [hunk 3]
```diff
@@ -28,6 +29,8 @@ class Message():
 
 	def __init__(self, **kwargs):
 		self.log = logging.getLogger(self.__class__.__name__)
+		self.uuid = uuid.uuid4()
+		self.sync = False
```
Comment (richteer) [inline halibot/message.py:33]: Clever, I like this. Makes the separation a bit easier

File: halibot/halmodule.py [hunk 4]
```diff
@@ -4,18 +4,23 @@
...
-
 		# For deprecation, remove for 1.0
 		if 'context' in kwargs:
 			msg.context = kwargs['context']
 
-		self.send_to(msg, [ msg.origin ])
+		msg = Message(body=body, type=mtype, author=author, origin=origin)
+
+		# Synchronous reply?
+		if msg0.sync:
```
Comment (richteer) [inline halibot/halmodule.py:20]: Where is `msg0` set? Automatically in a `sync_` environment? Or does this need to be provided by the module implementation of `.receive()`?

**Review by richteer — COMMENTED** [review]
Looks pretty good, we can go over it more in the next meeting.

File: halibot/halmodule.py (hunk 4 above)
Comment (richteer) [reply halibot/halmodule.py:20]: Nevermind, I'm an idiot. Didn't see that `msg0` has been like that since forever.

**Comment by sjrct** [comment]
As per 2017-08-01 meeting, merging

**Pull request merged**